### PR TITLE
WIP: OSRD goes async

### LINKED ISSUE #6679: Add a custom core-controller to manage specialized core instances
### Description and goal

Develop the core-controller, which monitor requests from editoast in queues and orchestrates the startup of core instances as well as all the surrounding elements.

### Acceptance criteria

- [X] Observe the REST API of RabbitMQ to schedule pods
- [X] Local Docker Support: Integrate the core-controller with the Docker API to control the lifecycle of the containers.
- [X] Kubernetes Support in Production: Develop the integration between the core-controller and the Kubernetes API to manage the lifecycle of the core-controller.
- [X] Using Redis to Manage the Controller's State: use Redis to store the state of the Core Controller. This is to ensure that the Core Controller can be horizontally scalable.
- [X] API Implementation: Implement an API through which either the gateway directly, or editoast, can inquire about the state of the different cores.

### PR DESCRIPTION
**Pull request by ghost**
This PR holds the whole modifications for making OSRD async using a RabbitMQ. This is a work in progress, here is an overview:

- [X] Modify local environement to provide rabbitmq
- [X] Implement Osrdyne
  - [X] Implement Kubernetes driver
  - [X] Implement Docker driver
  - [X] Implement RabbitMQ driver
  - [X] Implement logic
- [X] Test the whole stack
  - [X] Against Docker
  - [X] Against Kubernetes
- [X] Modify Core to use queues
  - [X] Create a client library for enqueing/dequeing messages
  - [X] Use the said library
- [X] Modify Editoast to use queues
  - [X] Create a client library for enqueing/dequeing messages
  - [X] Use the said library

Closes #6679, #6680, #6681

Helm cart related PR : https://github.com/OpenRailAssociation/osrd-chart/pull/11

### CODE REVIEW

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -25,6 +25,9 @@
 # --- CORE ---
 /core                                                    @osrd-project/core-maintainers
 
+# --- CORE CONTROLLER ---
+/core_controller                                         @osrd-project/devops
```
Comment (Khoyo) [inline .github/CODEOWNERS]: TODO: sync this with the openrail migration

File: core_controller/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+tokio = { version = "1.36", features = ["full"] }
+k8s-openapi = { version = "0.21", features = ["v1_29"] }
+serde = { version = "1", features = ["derive"] }
+reqwest = { version = "0.12", features = ["json"] }
+serde_json = "1.0"
+tracing = "0.1"
+tracing-subscriber = { version = "0.3", features = ["env-filter"] }
+redis = { version = "0.25", features = ["tokio-comp"] }
+chrono = "0.4"
+axum = { version = "0.7", features = ["macros"] }
```
Comment (flomonster) [inline core_controller/Cargo.toml]: Why using `axum` instead of `actix`?

File: core_controller/src/config.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+impl Default for Config {
+    fn default() -> Config {
+        // if env DEFAULT_CC_RMQ_HOST is not set, use rabbitmq (container name) otherwise use the env value
+        // Used for local development, making easier to seperate host mode to docker network mode
+        let default_rmq_host = match std::env::var("DEFAULT_CC_RMQ_HOST") {
+            Ok(val) => val,
+            Err(_) => "rabbitmq".to_string(),
+        };
```
Comment (flomonster) [inline core_controller/src/config.rs]:
```suggestion
        let default_rmq_host =
            std::env::var("DEFAULT_CC_RMQ_HOST").unwrap_or("rabbitmq".to_string());
```

File: core_controller/src/control_loop.rs [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+        };
+
+        let mut remaining_cores: Vec<CoreMetadata> = vec![];
+
+        // For each core, check when the last message was seen on redis and if it is older than the timeout, delete the core and its queue
+        for core in cores {
+            let last_seen: Option<i64> = redis::cmd("GET")
+                .arg(format!(
+                    "{}/{}",
+                    &self.config.redis.core_last_seen_prefix, core.infra_id
```
Comment (flomonster) [inline core_controller/src/control_loop.rs]: I'm suggesting to use `.` for consistency with editoast redis keys.
```suggestion
                    "{}.{}",
                    &self.config.redis.core_last_seen_prefix, core.infra_id
```

**Review by flomonster — COMMENTED** [review]
I haven't reviewed everything yet, but here's the first batch of comments. The PR is really good :+1:

File: core_controller/Cargo.toml (hunk 2 above)
Comment (ghost) [reply core_controller/Cargo.toml]: I lost my mind trying to wrap it inside the tokio::select!, with state and concurrency, and I found a solution with axum quite easily, so I kept it like this.

I can try to integrate actix again now that I've finished the State stuff etc... see if it works properly. Let me know if you think we should do this.

Comment (flomonster) [reply core_controller/Cargo.toml]: I see, no need to change it now.

**Review by flomonster — COMMENTED** [review]
Thanks for this PR! It looks good to me. 

> [!IMPORTANT]
> Can you run cargo fmt with the following format config: https://github.com/OpenRailAssociation/osrd/pull/7049

> [!NOTE]
> It's difficult for me to review the `kubernetes.rs` file since I'm not familiar with how k8s works.

File: editoast/src/core/mq_client.rs [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+
+        Ok(())
+    }
+
+    pub async fn call_with_response<T, TR>(
+        &self,
+        message_type: &str,
+        infra_id: i64,
+        infra_expected_version: i64,
+        payload: T,
```
Comment (multun) [inline editoast/src/core/mq_client.rs]: why not just take a slice of bytes, return a slice of bytes, and let the caller deal with it?

File: editoast/src/core/mq_client.rs [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct MessageEnvelope<T>
+where
+    T: Serialize + Debug,
+{
+    payload: T,
+    message_type: String,
+    respond_to: Option<String>,
```
Comment (multun) [inline editoast/src/core/mq_client.rs]: I'd rather move the respond_to to amqp headers. it has several upsides:
- the worker can still respond even if it cannot deserialize the message
- if `message_type` and `infra_expected_version` are moved out of this layer, which I don't see any downside to, RPC messages are just `[u8]`. the business of message format and parsing can be moved out of this layer

File: editoast/src/core/mq_client.rs [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+        let core_queue_name = self.create_queue_for_core(&channel, infra_id).await?;
+
+        // Publish the message to the core queue
+        channel
+            .basic_publish(
+                &self.exchange,
+                core_queue_name.as_str(),
+                BasicPublishOptions::default(),
+                &serialized_payload,
+                BasicProperties::default(),
```
Comment (multun) [inline editoast/src/core/mq_client.rs]: https://docs.rs/lapin/2.3.4/lapin/protocol/basic/struct.AMQPProperties.html#method.with_reply_to

We could also use https://www.rabbitmq.com/docs/direct-reply-to

File: core_controller/src/control_loop.rs [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+                        );
+                        return Err("Failed to delete queue".into());
+                    } else {
+                        info!("Deleted queue for infra_id: {}", core.infra_id);
+                    }
+
+                    // Remove the redis key
+                    if let Err(e) = redis::cmd("DEL")
+                        .arg(format!(
+                            "{}/{}",
```
Comment (multun) [inline core_controller/src/control_loop.rs]:
```suggestion
                            "{}.{}",
```

File: core_controller/src/config.rs [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        let default_redis_host: String =
+            std::env::var("DEFAULT_CC_REDIS_HOST").unwrap_or("redis".to_string());
+
+        Config {
+            loop_interval: 3,
+            core_timeout: 900,
+            api_listen_addr: "0.0.0.0:6000".to_string(),
+            redis: RedisOptions {
+                url: format!("redis://{}:6379/0", default_redis_host),
+                core_last_seen_prefix: "core/last_seen_msg".to_string(),
```
Comment (multun) [inline core_controller/src/config.rs]:
```suggestion
                core_last_seen_prefix: "core.last_seen_msg".to_string(),
```

**Review by multun — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
Tested Kubernetes driver against a real kubernetes cluster deployed for the tests. Worked properly.

**Comment by ghost** [comment]
Since I started to work on this PR, we've revamped at multiple occasion the architecture of the scalable Async RPC. We ended up on something really different that what was first envisioned.

Closing this PR in favor of a one against with a new branch.

**Pull request closed without merging**